src/Nethermind/Nethermind.Merge.Plugin/InvalidChainTracker/InvalidChainTracker.cs

```diff
@@ -0,0 +1,188 @@
+    private readonly LruCache<Keccak, Node> _tree;
+    
+    // CompositeDisposable only available on System.Reactive. So this will do for now. 
+    private readonly List<Action> _disposables = new();
+
+    public InvalidChainTracker(
+        IPoSSwitcher poSSwitcher, 
+        IBlockFinder blockFinder,
+        IBlockCacheService blockCacheService,
+        ILogManager logManager)
+    {
+        _poSSwitcher = poSSwitcher;
+        _blockFinder = blockFinder;
+        _tree = new LruCache<Keccak, Node>(1024, nameof(InvalidChainTracker));
+        _logger = logManager.GetClassLogger<InvalidChainTracker>();
+        _blockCacheService = blockCacheService;
+    }
+
+    public void SetupBlockchainProcessorInterceptor(IBlockchainProcessor blockchainProcessor)
+    {
+        blockchainProcessor.OnInvalidBlock += BlockchainProcessorOnOnInvalidBlock;
+        _disposables.Add(() =>
+        {
+            blockchainProcessor.OnInvalidBlock -= BlockchainProcessorOnOnInvalidBlock;
+        });
+    }
+
+    private void BlockchainProcessorOnOnInvalidBlock(object? sender, InvalidBlockException e)
+    {
+        OnInvalidBlock(e.InvalidBlockHash, null);
+    }
+
+    public void SetChildParent(Keccak child, Keccak parent)
+    {
+        lock (_opLock)
+        {
+            Node parentNode = GetNode(parent);
+
+            parentNode.Children.Add(child);
+            if (parentNode.LastValidHash != null)
+            {
+                PropagateLastValidHash(parentNode);
+            }
+        }
+    }
+
+    private Node GetNode(Keccak hash)
+    {
+        if (!_tree.TryGet(hash, out Node node))
+        {
+            node = new Node();
+            _tree.Set(hash, node);
+        }
+
+        return node;
+    }
+
+    private void PropagateLastValidHash(Node node)
+    {
+        foreach (Keccak nodeChild in node.Children)
+        {
+            Node childNode = GetNode(nodeChild);
+            if (childNode.LastValidHash != node.LastValidHash)
+            {
+                childNode.LastValidHash = node.LastValidHash;
+                PropagateLastValidHash(childNode);
+            }
+        }
+    }
+
+    private BlockHeader? TryGetBlockHeader(Keccak hash)
+    {
+        if (_blockCacheService.BlockCache.TryGetValue(hash, out Block block))
+        {
+            return block.Header;
+        }
+
+        return _blockFinder.FindHeader(hash);
+    }
+
+    public void OnInvalidBlock(Keccak failedBlock, Keccak? parent)
+    {
+        if (parent == null)
+        {
+            BlockHeader? failedBlockHeader = TryGetBlockHeader(failedBlock);
+            if (failedBlockHeader == null)
+            {
+                if(_logger.IsWarn) _logger.Warn($"Unable to resolve block to determine parent. Block {failedBlock}");
+                return;
+            }
+
+            parent = failedBlockHeader.ParentHash;
+        }
+        
+        Keccak effectiveParent = parent;
+        BlockHeader? parentHeader = TryGetBlockHeader(parent);
```

Comment:
just take in Headers in first place, they should be available in all scenarios: `OnInvalidBlock(BlockHeader failedBlock, BlockHeader parent)`

Reply:
Ok

Reply:
Problem, sometime it does not have the parent header. Eg, in case of newPayload where it's parent is not synced or missed.

Reply:
ok, but maybe lets try resolve parent upstream then?

Reply:
Upstream?

Reply:
Upostream - the code that is calling this code.